From a1a8923430da4f1f8fa24e9b0ba401b2f14f277a Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Wed, 5 Nov 2014 17:40:41 -0500
Subject: [ethernet] be2net: use v1 of SET_FLOW_CONTROL command

Message-id: <1415209241-20744-1-git-send-email-ivecera@redhat.com>
Patchwork-id: 99408
O-Subject: [RHEL7.1 PATCH] be2net: use v1 of SET_FLOW_CONTROL command
Bugzilla: 1087128
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: #1087128 - 'ethtool -A be2net rx off' return zero, but rx remains on

Description:
The patch fixes a bug described below.

Test status:
Tested by Emulex

Upstream commit(s):
commit b29812c13514b3bb9236aed4bd35192e6b3d0f2d
Author: Suresh Reddy <Suresh.Reddy@emulex.com>
Date:   Fri Sep 12 17:39:17 2014 +0530

    be2net: use v1 of SET_FLOW_CONTROL command

    In some configurations the FW doesn't allow changing flow control settings
    of a link. Unless a v1 version of the SET_FLOW_CONTROL cmd is used, the FW
    doesn't report an error to the driver.

    Signed-off-by: Suresh Reddy <Suresh.Reddy@emulex.com>
    Signed-off-by: Sathya Perla <sathya.perla@emulex.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/emulex/benet/be_cmds.c b/drivers/net/ethernet/emulex/benet/be_cmds.c
index 5be100d..7d496d4 100644
--- a/drivers/net/ethernet/emulex/benet/be_cmds.c
+++ b/drivers/net/ethernet/emulex/benet/be_cmds.c
@@ -1948,6 +1948,7 @@ int be_cmd_set_flow_control(struct be_adapter *adapter, u32 tx_fc, u32 rx_fc)
           OPCODE_COMMON_SET_FLOW_CONTROL, sizeof(*req),
           wrb, NULL);
 
+ req->hdr.version = 1;
  req->tx_flow_control = cpu_to_le16((u16)tx_fc);
  req->rx_flow_control = cpu_to_le16((u16)rx_fc);
 
@@ -1955,6 +1956,10 @@ int be_cmd_set_flow_control(struct be_adapter *adapter, u32 tx_fc, u32 rx_fc)
 
 err:
  spin_unlock_bh(&adapter->mcc_lock);
+
+ if (base_status(status) == MCC_STATUS_FEATURE_NOT_SUPPORTED)
+  return  -EOPNOTSUPP;
+
  return status;
 }
 
diff --git a/drivers/net/ethernet/emulex/benet/be_cmds.h b/drivers/net/ethernet/emulex/benet/be_cmds.h
index 0e11868..3b1606c 100644
--- a/drivers/net/ethernet/emulex/benet/be_cmds.h
+++ b/drivers/net/ethernet/emulex/benet/be_cmds.h
@@ -57,7 +57,8 @@ enum mcc_base_status {
  MCC_STATUS_ILLEGAL_FIELD = 3,
  MCC_STATUS_INSUFFICIENT_BUFFER = 4,
  MCC_STATUS_UNAUTHORIZED_REQUEST = 5,
- MCC_STATUS_NOT_SUPPORTED = 66
+ MCC_STATUS_NOT_SUPPORTED = 66,
+ MCC_STATUS_FEATURE_NOT_SUPPORTED = 68
 };
 
 /* Additional status */
-- 
1.7.1